File: frontend/machsense/lib/api/notifications.ts
import { api } from "./client";

export interface NotificationOut {
  id: number;
  recipient_id: number;
  machine_id?: number;
  alert_id?: number;
  maintenance_id?: number;
  type: string;
  title: string;
  message?: string;
  is_read: boolean;
  created_at?: string;
}

export const notificationsApi = {
  list: (unread_only = false) =>
    api.get<NotificationOut[]>(`/notifications/?unread_only=${unread_only}`),
  markRead: (id: number) => api.patch<{ message: string }>(`/notifications/${id}/read`),
  markAllRead: () => api.post<{ message: string }>("/notifications/read-all"),
};

File: backend/app/routers/notifications.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List

from app.db.database import get_db
from app.db.models.notification import Notification
from app.db.models.user import User
from app.schemas.schemas import NotificationOut
from app.core.security import get_current_user
from app.services.notification_service import get_notifications, mark_read, mark_all_read

router = APIRouter(prefix="/notifications", tags=["Notifications"])


@router.get("/", response_model=List[NotificationOut])
async def list_notifications(
    unread_only: bool = False,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return get_notifications(db, current_user.id, unread_only)


@router.patch("/{notification_id}/read", status_code=200)
async def mark_notification_read(
    notification_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    success = mark_read(db, notification_id, current_user.id)
    if not success:
        raise HTTPException(status_code=404, detail="Notification not found")
    return {"message": "Marked as read"}


@router.post("/read-all", status_code=200)
async def mark_all_notifications_read(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    count = mark_all_read(db, current_user.id)
    return {"message": f"{count} notifications marked as read"}
